presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java

```diff
@@ -242,18 +247,62 @@ public static long fromISO8601Date(ConnectorSession session, @SqlType("varchar(x
+    public static long timeAtTimeZone(ConnectorSession session, @SqlType(StandardTypes.TIME_WITH_TIME_ZONE) long timeWithTimeZone, @SqlType(StandardTypes.INTERVAL_DAY_TO_SECOND) long zoneOffset)
     {
         checkCondition((zoneOffset % 60_000) == 0, INVALID_FUNCTION_ARGUMENT, "Invalid time zone offset interval: interval contains seconds");
         int zoneOffsetMinutes = (int) (zoneOffset / 60_000);
-        return packDateTimeWithZone(unpackMillisUtc(timeWithTimeZone), getTimeZoneKeyForOffset(zoneOffsetMinutes));
+        return timeAtTimeZone(session, timeWithTimeZone, getTimeZoneKeyForOffset(zoneOffsetMinutes));
+    }
+
+    private static long timeAtTimeZone(ConnectorSession session, long timeWithTimeZone, TimeZoneKey timeZoneKey)
+    {
+        DateTimeZone sourceTimeZone = getDateTimeZone(unpackZoneKey(timeWithTimeZone));
+        DateTimeZone targetTimeZone = getDateTimeZone(timeZoneKey);
+        long millis = unpackMillisUtc(timeWithTimeZone);
+
+        // STEP 1. Calculate source UTC millis in session start
+        millis += valueToSessionTimeZoneOffsetDiff(millis, session.getStartTime(), sourceTimeZone);
+
+        // STEP 2. Calculate target UTC millis in 1970
+        millis -= valueToSessionTimeZoneOffsetDiff(millis, session.getStartTime(), targetTimeZone);
+
+        // STEP 3. Make sure that value + offset is in 0 - 23:59:59.999
+        long localMillis = millis + targetTimeZone.getOffset(0);
+        // Loops up to 2 times in total
+        while (localMillis > TimeUnit.DAYS.toMillis(1)) {
+            millis -= TimeUnit.DAYS.toMillis(1);
+            localMillis -= TimeUnit.DAYS.toMillis(1);
+        }
+        while (localMillis < 0) {
+            millis += TimeUnit.DAYS.toMillis(1);
+            localMillis += TimeUnit.DAYS.toMillis(1);
+        }
+
+        return packDateTimeWithZone(millis, timeZoneKey);
+    }
+
+    // HACK WARNING!
+    // This method does calculate difference between timezone offset in current date (session start)
+    // and 1970-01-01 (same timezone). This is used to be able to avoid using fixed offset TZ id for
+    // formatter.
+    // DateTimeFormatter does format specified instance in specified time zone calculating offset for
+    // that time zone based on provided instance. As Presto TIME type is represented as millis since
+    // 00:00.000 of some day UTC, we always use timezone offset that was valid on 1970-01-01.
+    // Best effort without changing representation of TIME WITH TIME ZONE is to use offset of the timezone
+    // based on session start time.
+    // By adding this difference to instance that we would like to format, we can trick formetter to
+    // applying part of offset that was valid on 1970-01-01 and we apply "the rest" manually.
+    // Silent assumption is made, that no changes in TZ offsets were done on 1970-01-01.
```

Comment:
ya ya I just wanted to keep it as short as possible.
It's long and complicated anyway. Getting into to much details doesn't help.